blockchain/types/src/avr_history.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Types for reading and writing historical Attestation Verification
+//! Reports from MobileCoin consensus enclaves
+
+use alloc::{string::String, vec::Vec};
+use hex::ToHex;
+use mc_attest_core::VerificationReport;
+use mc_crypto_keys::Ed25519Public;
+use mc_common::HashMap;
+use toml::{map::Map, value::Value};
+use url::Url;
+
+/// Object representing AVR history for all consensus nodes
+pub struct AVRHistoricalRecords {
+    /// AVR Records
+    pub avr_records: HashMap<Url, Vec<AVRHistoricalRecord>>,
+}
+
+impl AVRHistoricalRecords {
+    /// Create new records object
+    pub fn new(avr_records: HashMap<Url, Vec<AVRHistoricalRecord>>) -> Self {
+        Self { avr_records }
+    }
+
+    /// Get a toml representation of the AVR historical records
+    pub fn export_toml_string(&self) -> Result<String, toml::ser::Error>{
+        toml::to_string(&Map::from(self))
+    }
+}
+
+/// Stores a single historical AVR for each tuple of responder_id
+/// & block signing key  in the format specified in draft MCIP #43
+pub struct AVRHistoricalRecord {
+    /// Uri of the consensus node
+    pub node_url: Url,
+
+    /// Block signing key of the consensus enclave
+    pub public_key: Option<Ed25519Public>,
+
+    /// Block the AVR for the signing key becomes valid
+    pub start_block: u64,
+
+    /// Final block the AVR for the signing key is valid
+    pub end_block: u64,
+
+    /// Hex of the AVR bytes
+    pub avr: Option<VerificationReport>,
```

Comment:
It seems better to make this required; an AVR record doesn't really make sense without an AVR...

Reply:
There are cases where there aren't AVRs available

Reply:
Then we shouldn't emit a record for those ranges ;)

Reply:
I think it's better to be explicit when deserializing the toml which records don't have ranges. "Missing AVR" is a valid answer.

Reply:
This was ultimately solved by the intermediate serialization format. For the pre-verifier config struct, the option should remain to indicate that some ranges didn't have an AVR present